physicsnemo/mesh/mesh.py

```diff
@@ -0,0 +1,1047 @@
+                    f"Current codimension: {self.codimension}"
+                )
+
+            ### Get cell normals and areas (triggers computation if not cached)
+            cell_normals = self.cell_normals  # (n_cells, n_spatial_dims)
+            cell_areas = self.cell_areas  # (n_cells,)
+
+            ### Initialize accumulated weighted normals for each point
+            # Shape: (n_points, n_spatial_dims)
+            weighted_normals = torch.zeros(
+                (self.n_points, self.n_spatial_dims),
+                dtype=self.points.dtype,
+                device=self.points.device,
+            )
+
+            ### Vectorized accumulation of area-weighted normals
+            # For each cell, add (cell_normal * cell_area) to each of its vertices
+
+            # Get all vertex indices from all cells
+            # Shape: (n_cells, n_vertices_per_cell)
+            n_vertices_per_cell = self.cells.shape[1]
+
+            # Flatten point indices: (n_cells * n_vertices_per_cell,)
+            point_indices = self.cells.flatten()
+
+            # Repeat cell normals for each vertex in the cell
+            # Shape: (n_cells, n_vertices_per_cell, n_spatial_dims)
+            cell_normals_repeated = cell_normals.unsqueeze(1).expand(
+                -1, n_vertices_per_cell, -1
+            )
+            # Flatten: (n_cells * n_vertices_per_cell, n_spatial_dims)
+            cell_normals_flat = cell_normals_repeated.reshape(-1, self.n_spatial_dims)
+
+            # Repeat cell areas for each vertex in the cell
+            # Shape: (n_cells, n_vertices_per_cell)
+            cell_areas_repeated = cell_areas.unsqueeze(1).expand(
+                -1, n_vertices_per_cell
+            )
+            # Flatten: (n_cells * n_vertices_per_cell,)
+            cell_areas_flat = cell_areas_repeated.flatten()
+
+            # Weight normals by area
+            # Shape: (n_cells * n_vertices_per_cell, n_spatial_dims)
+            weighted_normals_flat = cell_normals_flat * cell_areas_flat.unsqueeze(-1)
+
+            ### Scatter-add weighted normals to their corresponding points
+            # Expand point_indices to match weighted_normals_flat shape
+            point_indices_expanded = point_indices.unsqueeze(-1).expand(
+                -1, self.n_spatial_dims
+            )
+
+            # Accumulate weighted normals at each point
+            weighted_normals.scatter_add_(
+                dim=0,
+                index=point_indices_expanded,
+                src=weighted_normals_flat,
+            )
+
+            ### Normalize to get unit normals
+            # For isolated points (zero weighted sum), F.normalize returns zero vector
+            cached = F.normalize(weighted_normals, dim=-1)
+            set_cached(self.point_data, "normals", cached)
+
+        return cached
+
+    @classmethod
+    def merge(
+        cls, meshes: Sequence["Mesh"], global_data_strategy: Literal["stack"] = "stack"
```

Comment:
do all meshes need to be on the same device here?

Reply:
Fixed